lib/linked_list.rb

```diff
@@ -6,158 +6,294 @@ class Node
 
   def initialize(value, next_node = nil)
     @data = value
-    @next = next_node
+    @next = nil
   end
 end
 
 # Defines the singly linked list
 class LinkedList
-    def initialize
-      @head = nil # keep the head private. Not accessible outside this class
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+  end
+
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  # Time Complexity: Constant
+  # Space Complexity: Constant
+  def add_first(value)
+    new_node = Node.new(value)
+
+    if @head != nil #if list isn't empty
+      new_node.next = @head
     end
 
-    # method to add a new node with the specific data value in the linked list
-    # insert the new node at the beginning of the linked list
-    # Time Complexity:  
-    # Space Complexity
-    def add_first(value)
-      raise NotImplementedError
+    @head = new_node
+  end
+
+  # method to find if the linked list contains a node with specified value
+  # returns true if found, false otherwise
+  # Time Complexity: o(n) where n is the length of the linked list because this method iterates through each node to find the value
+  # Space Complexity: constant, no extra space needed
+  def search(value)
+    if @head == nil
+      return false
     end
 
-    # method to find if the linked list contains a node with specified value
-    # returns true if found, false otherwise
-    # Time Complexity:  
-    # Space Complexity
-    def search(value)
-      raise NotImplementedError
+    current = @head
+    while @next != nil

```

Comment:
Don't you mean `while current != nil`, and you should be doing `current = current.next`